### ISSUE TITLE
[Feature][3.4] Set initial CrudPanel properties on CrudController

### ISSUE THREAD

**Issue by tabacitu** [Laravel-Backpack/CRUD#1245]
WHAT IF we provided ```EntityCrudController``` with some properties that the user could easily overwrite, to achieve the same functionality as ```$this->crud->enableExportButtons()``` or similar calls? 

Basically to set CrudPanel properties directly on the CrudController, as properties. Ugh. Difficult to explain. Here's how I _think_ an EntityCrudController could look like:

```diff
<?php

namespace Backpack\MenuCRUD\app\Http\Controllers\Admin;

use App\Http\Requests;
use Backpack\CRUD\app\Http\Controllers\CrudController;

// VALIDATION: change the requests to match your own file names if you need form validation
use Backpack\CRUD\app\Http\Requests\CrudRequest as StoreRequest;
use Backpack\CRUD\app\Http\Requests\CrudRequest as UpdateRequest;

class MenuItemCrudController extends CrudController
{
+    private $model = "App\Models\MenuItem";
+    private $route = backpack_url('menu-item');
+    private $entity_name = "menu item";
+    private $entity_name_plural = "menu items";

+    private $reorder = true;
+    private $reorder_label = 'name';
+    private $reorder_max_level = 2;

+    private $details_row = false;
+    private $export_buttons = false;

    public function setup()
    {
-        $this->crud->setModel("Backpack\MenuCRUD\app\Models\MenuItem");
-        $this->crud->setRoute(config('backpack.base.route_prefix').'/menu-item');
-        $this->crud->setEntityNameStrings('menu item', 'menu items');

-        $this->crud->allowAccess('reorder');
-        $this->crud->enableReorder('name', 2);

        $this->crud->addColumn([
                                'name' => 'name',
                                'label' => 'Label',
                            ]);
        $this->crud->addColumn([
                                'label' => 'Parent',
                                'type' => 'select',
                                'name' => 'parent_id',
                                'entity' => 'parent',
                                'attribute' => 'name',
                                'model' => "\Backpack\MenuCRUD\app\Models\MenuItem",
                            ]);

        $this->crud->addField([
                                'name' => 'name',
                                'label' => 'Label',
                            ]);
        $this->crud->addField([
                                'label' => 'Parent',
                                'type' => 'select',
                                'name' => 'parent_id',
                                'entity' => 'parent',
                                'attribute' => 'name',
                                'model' => "\Backpack\MenuCRUD\app\Models\MenuItem",
                            ]);
        $this->crud->addField([
                                'name' => 'type',
                                'label' => 'Type',
                                'type' => 'page_or_link',
                                'page_model' => '\Backpack\PageManager\app\Models\Page',
                            ]);
    }

    public function store(StoreRequest $request)
    {
        return parent::storeCrud($request);
    }

    public function update(UpdateRequest $request)
    {
        return parent::updateCrud($request);
    }
}

```

PROs:
- onboarding - easier to get to know Backpack features exist (since the property is there in every CrudController you generate;
- onboarding - a little easier to toggle on/off some functionality, since the code is already there, just change ```false``` into ```true```;
- it makes more sense from a OOP perspective (those properties are used controller-wide, so setting them in the ```setup()``` method is a bit deceiving); then again, the ```setup()``` method would continue to set variables controller-wide for fields, columns, filters, etc;
- you end up with a cleaner ```setup()``` method;

CONs:
- if both options exist (property on CrudController and CrudPanel method), it could become difficult to understand which one takes precedence (it's obvious to me that ```$this->crud->setModel('App\Models\Monster');``` would overwrite the ```$model``` property you previously set, but I don't know if it would be obvious for everyone;
- basically ```CrudController``` would create a ```CrudPanel``` object with whatever properties are set on it; so we'll end up with the same properties both on the ```CrudController``` object, and on the ```CrudPanel``` object; not very DRY; we could just transfer them using ```get_object_vars()``` but I'm not 100% sure that will work; plus if someone uses custom properties on a custom EntityCrudController, those would get transfered too;

Thoughts?

**Comment by tabacitu** [comment]
@Laravel-Backpack/veterans what do you guys think?

**Comment by OwenMelbz** [comment]
I sort of like, but also sort of dont like.

**Pro**
I like because sure it can reduce clutter.

**Cons**
I dislike because....

1. 
2. stuff is then stored in 2 places, using different style, 1/2 fluent syntax, 1/2 props
3. logic is a bit more obscure if you want to wrap things in logic, e.g only certain users can export things
4. I always define things such as `setModel(User::class)` which seems to becoming a more common pattern in laravel, That wouldn't work using this style?
5. Can you actually use "backpack_url" helper at the top?
6. Would it make sense to make them constants rather than just props?
7. It then introduces code styling/standard issues, e.g if your project enforces camelCaseStyleVariables, you're then forcing users to use underscore_style_variables - which could be annoying and cause issues with linters
8. Can you set the entity translation strings from up there? for the single/plurals?

**Comment by adriandmitroca** [comment]
@OwenMelbz 

4&5: perhaps as alternative dynamic values could be set within the constructor? that would do the job (though this would be third place where you setup stuff)

**Comment by tabacitu** [comment]
@adriandmitroca - yup, the constructor would be a third place to do that... so even more complicated... But you COULD, theoretically, overwrite anything you want in the ```setup()``` method, like you do now. Anything you do in ```setup()``` would basically overwrite the props. So no need for a third place. Just the two.

@OwenMelbz Honestly, I like your PROs and CONs better than mine :-) I think they're more objective. It's obvious to me now that I was biased when writing mine - so thanks a lot for the feedback! Ok... now I'm leaning towards NOT implementing this. It seems like a change that would bring a small benefit (cleaner code), but at the detriment of making things more difficult to customize. Simpler to just have the one way to do stuff, that allows you maximum customization.

I'll leave this open for a little longer, in case someone else wants to pitch in with their opinion. As you can see, I'm easily swayed. But if this isn't closed in 60 days, I think we should let Justin close it :-)
